api/model/form/location.go

```diff
@@ -0,0 +1,524 @@
+package form
+
+import (
+	"encoding/json"
+	"log"
+	"strings"
+
+	"github.com/18F/e-QIP-prototype/api/geo"
+	"github.com/18F/e-QIP-prototype/api/model"
+)
+
+// Different potential layouts used by the frontend.
+// File: /src/components/Form/Location/Layouts.js
+const (
+	LayoutAddress                             = "Address"
+	LayoutBirthPlace                          = "Birthplace"
+	LayoutBirthPlaceWithoutCounty             = "Birthplace without County"
+	LayoutCountry                             = "Country"
+	LayoutUSCityStateInternationalCity        = "US City, State, International city"
+	LayoutUSCityStateInternationalCityCountry = "US City, State, International city country"
+	LayoutCityState                           = "City, State"
+	LayoutStreetCityCountry                   = "Street, City, Country"
+	LayoutCityCountry                         = "City, Country"
+	LayoutUSCityStateZipcodeInternationalCity = "US City, State, Zipcode International city"
+	LayoutCityStateCountry                    = "City, State, Country"
+	LayoutUSAddress                           = "US Address"
+	LayoutStreetCity                          = "Street, City"
+)
+
+// Location is a basic input.
+type Location struct {
+	Layout    string `json:"layout"`
+	Street1   string `json:"street,omitempty"`
+	Street2   string `json:"street2,omitempty"`
+	City      string `json:"city,omitempty"`
+	State     string `json:"state,omitempty"`
+	Zipcode   string `json:"zipcode,omitempty"`
+	County    string `json:"county,omitempty"`
+	Country   string `json:"country,omitempty"`
+	Validated bool   `json:"validated,omitempty"`
+}
+
+// Unmarshal bytes in to the entity properties.
+func (entity *Location) Unmarshal(raw []byte) error {
+	return json.Unmarshal(raw, entity)
+}
+
+// Valid checks the value(s) against an battery of tests.
+func (entity *Location) Valid() (bool, error) {
+	log.Printf("location: %v", entity)
+	if entity.Validated {
+		return true, nil
+	}
+
+	var stack model.ErrorStack
+	domestic := entity.Country == "United States"
+	postoffice := entity.Country == "POSTOFFICE"
+	international := !domestic && !postoffice
+
+	switch entity.Layout {
+	case LayoutBirthPlace:
+		if domestic {
+			stack = validateFields(entity, "city", "state", "county")
+		} else {
+			stack = validateFields(entity, "city", "county")
+		}
+	case LayoutBirthPlaceWithoutCounty:
+		if domestic {
+			stack = validateFields(entity, "city", "state")
+		} else {
+			stack = validateFields(entity, "city", "county")
+		}
+	case LayoutCountry:
+		stack = validateFields(entity, "country")
+	case LayoutUSCityStateInternationalCity:
+		if domestic {
```

Comment:
this if/else pattern and the `validateFields` lines get repeated a lot. maybe a good candidate to refactor.